File: apps/mac/Sources/Design/RunVisuals.swift
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
//
// Source-available for review, NOT open source. See LICENSE: no rights to
// redistribute, publish, or ship a build are granted. Read it, study it, run
// your own build of it.
// "tokenstat" is a trademark of pueev OU. See TRADEMARK.md.

import SwiftUI

/// What a run's status looks like, in one place.
///
/// Automations and Workflows each carried their own copy of this switch, which
/// is how "stopped" ends up amber on one screen and grey on the next. The two
/// entry points that used to own it now call through here.
enum RunOutcome {
    static func tint(_ status: String) -> Color {
        switch status {
        case "running", "queued": return Theme.stateWorking
        case "waiting": return Theme.warning
        case "ok": return Theme.success
        case "stopped": return Theme.warning
        case "error": return Theme.danger
        case "interrupted": return Theme.warning
        default: return Theme.stateIdle
        }
    }
}

/// The last handful of outcomes as ticks, newest on the right.
///
/// A job whose last three runs failed should look wrong from across the room.
/// Before this, that fact lived in a shared "Recent runs" list at the bottom of
/// the screen, detached from the job it belonged to, so nobody read it as a
/// trend.
struct RunHistoryStrip: View {
    struct Tick: Identifiable {
        var id: String
        var status: String
        var label: String
    }

    /// Oldest first. The view reverses nothing, it just trims the front.
    var ticks: [Tick]
    var limit: Int = 12
    var height: CGFloat = 14
    var width: CGFloat = 4
    var onSelect: ((Tick) -> Void)?

    /// Slots always drawn, so the strip reads as a track that is filling up.
    ///
    /// Two amber ticks on their own are a pause glyph, and next to a Run button
    /// that is exactly how they were read. Empty slots behind them give the
    /// filled ones something to be part of.
    private static let slots = 8

    private var shown: [Tick] {
        ticks.count > limit ? Array(ticks.suffix(limit)) : ticks
    }

    var body: some View {
        HStack(spacing: 2) {
            ForEach(0..<max(0, Self.slots - shown.count), id: \.self) { _ in
                Capsule()
                    .fill(Theme.border.opacity(0.5))
                    .frame(width: width, height: height)
            }
            ForEach(shown) { tick in
                Capsule()
                    .fill(RunOutcome.tint(tick.status))
                    .frame(width: width, height: height)
                    .opacity(tick.status == "ok" ? 0.85 : 1)
                    .help(tick.label)
                    .onTapGesture { onSelect?(tick) }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(summary)
    }

    private var summary: String {
        if shown.isEmpty { return "Never run" }
        let failed = shown.filter { $0.status == "error" }.count
        if failed == 0 { return "Last \(shown.count) runs, none failed" }
        return "Last \(shown.count) runs, \(failed) failed"
    }
}

/// How long a run took, relative to the longest one beside it.
///
/// A forty-minute run and a forty-second one were the same row. The bar is
/// deliberately relative rather than absolute: the question a list of runs
/// answers is "which of these was the long one", not "how many seconds".
struct DurationBar: View {
    var seconds: Double
    var longest: Double
    var width: CGFloat = 44
    var height: CGFloat = 4

    private var fraction: Double {
        guard longest > 0, seconds > 0 else { return 0 }
        return min(1, max(0.06, seconds / longest))
    }

    var body: some View {
        ZStack(alignment: .leading) {
            Capsule()
                .fill(Theme.border)
                .frame(width: width, height: height)
            Capsule()
                .fill(Theme.accent.opacity(0.7))
                .frame(width: width * fraction, height: height)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
        .help(label)
    }

    private var label: String {
        guard seconds > 0 else { return "Still running" }
        if seconds < 60 { return "Took \(Int(seconds))s" }
        if seconds < 3600 { return "Took \(Int(seconds / 60))m" }
        let hours = seconds / 3600
        return String(format: "Took %.1fh", hours)
    }
}

File: apps/android/app/src/main/java/ai/tokenstat/tokenstat/ui/marks/RunVisuals.kt
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
package ai.tokenstat.tokenstat.ui.marks

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import ai.tokenstat.tokenstat.ui.theme.LocalTsColors
import ai.tokenstat.tokenstat.ui.theme.TsColors
import java.util.Locale

/// What a run's status looks like, in one place. Port of `RunVisuals.swift`:
/// the Apple client's Automations and Workflows screens each carried their own
/// copy of this switch, which is how one status ended up two colours. Both
/// entry points now call through here, and so does this one.
object RunOutcome {
    fun tint(status: String, colors: TsColors): Color = when (status) {
        "running", "queued" -> colors.stateWorking
        "waiting" -> colors.warning
        "ok" -> colors.success
        "stopped" -> colors.warning
        "error" -> colors.danger
        "interrupted" -> colors.warning
        else -> colors.stateIdle
    }
}

/// One past run on the strip. Oldest first; the strip trims the front.
data class RunTick(val id: String, val status: String, val label: String)

/// Slots always drawn, so the strip reads as a track that is filling up.
/// Two amber ticks on their own are a pause glyph; empty slots behind them
/// give the filled ones something to be part of.
const val RUN_HISTORY_SLOTS = 8

/// The ticks actually shown: the newest `limit` at most. A bad limit shows an
/// empty strip rather than crashing, like the Apple client's `max(0, limit)`.
fun visibleRunTicks(ticks: List<RunTick>, limit: Int = RUN_HISTORY_SLOTS): List<RunTick> {
    val cap = maxOf(0, limit)
    return if (ticks.size > cap) ticks.takeLast(cap) else ticks
}

fun runStripSummary(shown: List<RunTick>): String {
    if (shown.isEmpty()) return "Never run"
    val failed = shown.count { it.status == "error" }
    if (failed == 0) return "Last ${shown.size} runs, none failed"
    return "Last ${shown.size} runs, $failed failed"
}

/// The last handful of outcomes as ticks, newest on the right. A job whose
/// last three runs failed should look wrong from across the room.
@Composable
fun RunHistoryStrip(
    ticks: List<RunTick>,
    modifier: Modifier = Modifier,
    limit: Int = RUN_HISTORY_SLOTS,
    tickWidth: Dp = 4.dp,
    tickHeight: Dp = 14.dp,
    onSelect: ((RunTick) -> Unit)? = null,
) {
    val colors = LocalTsColors.current
    val shown = visibleRunTicks(ticks, limit)
    val pill = RoundedCornerShape(50)
    Row(
        modifier.semantics { contentDescription = runStripSummary(shown) },
        horizontalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        repeat(maxOf(0, RUN_HISTORY_SLOTS - shown.size)) {
            Box(
                Modifier
                    .size(tickWidth, tickHeight)
                    .clip(pill)
                    .background(colors.border.copy(alpha = 0.5f)),
            )
        }
        shown.forEach { tick ->
            val tickModifier = Modifier
                .size(tickWidth, tickHeight)
                .clip(pill)
                .background(RunOutcome.tint(tick.status, colors).let {
                    if (tick.status == "ok") it.copy(alpha = 0.85f) else it
                })
            Box(
                if (onSelect != null) tickModifier.clickable { onSelect(tick) } else tickModifier,
            )
        }
    }
}

/// How long a run took, relative to the longest one beside it. Deliberately
/// relative rather than absolute: the question a list of runs answers is
/// "which of these was the long one", not "how many seconds".
fun durationFraction(seconds: Double, longest: Double): Double {
    if (longest <= 0 || seconds <= 0) return 0.0
    return minOf(1.0, maxOf(0.06, seconds / longest))
}

fun durationLabel(seconds: Double): String {
    if (seconds <= 0) return "Still running"
    if (seconds < 60) return "Took ${seconds.toInt()}s"
    if (seconds < 3600) return "Took ${(seconds / 60).toInt()}m"
    return "Took ${"%.1f".format(Locale.US, seconds / 3600)}h"
}

@Composable
fun DurationBar(
    seconds: Double,
    longest: Double,
    modifier: Modifier = Modifier,
    width: Dp = 44.dp,
    height: Dp = 4.dp,
) {
    val colors = LocalTsColors.current
    val pill = RoundedCornerShape(50)
    val label = durationLabel(seconds)
    Box(
        modifier
            .size(width, height)
            .semantics { contentDescription = label },
    ) {
        Box(
            Modifier
                .fillMaxSize()
                .clip(pill)
                .background(colors.border),
        )
        Box(
            Modifier
                .fillMaxHeight()
                .fillMaxWidth(durationFraction(seconds, longest).toFloat())
                .clip(pill)
                .background(colors.accent.copy(alpha = 0.7f)),
        )
    }
}
